sklearn/cluster/_hdbscan/hdbscan.py

```diff
@@ -0,0 +1,718 @@
+        memory=None,
+        n_jobs=4,
+        cluster_selection_method="eom",
+        allow_single_cluster=False,
+        metric_params=None,
+    ):
+        self.min_cluster_size = min_cluster_size
+        self.min_samples = min_samples
+        self.alpha = alpha
+        self.max_cluster_size = max_cluster_size
+        self.cluster_selection_epsilon = cluster_selection_epsilon
+        self.metric = metric
+        self.algorithm = algorithm
+        self.leaf_size = leaf_size
+        self.memory = memory
+        self.n_jobs = n_jobs
+        self.cluster_selection_method = cluster_selection_method
+        self.allow_single_cluster = allow_single_cluster
+        self.metric_params = metric_params
+
+    def fit(self, X, y=None):
+        """Perform HDBSCAN clustering from features or distance matrix.
+
+        Parameters
+        ----------
+        X : array or sparse (CSR) matrix of shape (n_samples, n_features), or \
+                array of shape (n_samples, n_samples)
+            A feature array, or array of distances between samples if
+            `metric='precomputed'`.
+
+        y : Ignored
+            Ignored.
+
+        Returns
+        -------
+        self : object
+            Returns self.
+        """
+        self._validate_params()
+        self._metric_params = self.metric_params or {}
+        if self.metric != "precomputed":
+            # Non-precomputed matrices may contain non-finite values.
+            X = self._validate_data(
+                X, accept_sparse="csr", force_all_finite=False, dtype=np.float64
+            )
+            self._raw_data = X
+
+            all_finite = (
+                np.all(np.isfinite(X.data)) if issparse(X) else np.all(np.isfinite(X))
+            )
+
+            if not all_finite:
+                # Pass only the purely finite indices into hdbscan
+                # We will later assign all non-finite points to the
+                # background-1 cluster
+                finite_index = get_finite_row_indices(X)
+                X = X[finite_index]
+                internal_to_raw = {x: y for x, y in enumerate(finite_index)}
+                outliers = list(set(range(X.shape[0])) - set(finite_index))
+        elif issparse(X):
+            # Handle sparse precomputed distance matrices separately
+            X = self._validate_data(
+                X,
+                accept_sparse="csr",
+                dtype=np.float64,
+            )
+        else:
+            # Only non-sparse, precomputed distance matrices are handled here
+            # and thereby allowed to contain numpy.inf for missing distances
+
+            # Perform data validation after removing infinite values (numpy.inf)
+            # from the given distance matrix.
+            X = np.asarray(X)
+            tmp = X.copy()
+            tmp[np.isinf(tmp)] = 1
+            self._validate_data(tmp, dtype=np.float64)
+
+        self.n_features_in_ = X.shape[1]
+        self._min_samples = (
```

Comment:
I don't see much value in exposing it publicly, since it is fairly obvious to determine based on whether or not `min_samples` was passed to the constructor. I'm inclined to leave it as-is unless you feel strongly about it.

Reply:
Let's keep it private for the moment. It will always be easy to expose it if needed.